### ISSUE TITLE
🎅 Playwright Santa 2025 🛷 

### ISSUE THREAD

**Issue by pavelfeldman** [microsoft/playwright#33955]
Tell us what you'd like to see in Playwright in the next year!

**Comment by Smrtnyk** [comment]
this https://github.com/microsoft/playwright/issues/22791

**Comment by Bilge** [comment]
Fix the bugs, bro.

**Comment by jaktestowac** [comment]
Projects and UI Mode improvement to make working with multiple browser/scopes/environments configurations easy:
https://github.com/microsoft/playwright/issues/23084

**Comment by sand4rt** [comment]
The removal of the experimental flag on component testing🤞

**Comment by scott-au** [comment]
[[Feature] automatically handle frames](https://github.com/microsoft/playwright/issues/12201)

**Comment by Manjunath224** [comment]
Full support for Component testing with examples and documentation. 

Help channel in discord meant only for component testing.

**Comment by chensce** [comment]
https://github.com/microsoft/playwright/issues/23700

**Comment by jozefizso** [comment]
Easy extensibility to add support for different apps which may be hosting Chromium or WebView2 web views.

Our use case is a Microsoft Office app extension which runs in the Office app and displays content in the taskpane using web view.

Simillary to Android support where we can run Anroid app and connect to its web view, we would like to create integration between Playwright and the Office apps. This would allow us to test if the actions in the web view had effect on the document in the Office app.

We know how to implement the Office app COM addins would drive the automation in a similar way the DevTools protocol does. It would expose the application model and data to PW tests. We are stuggling with integrating this to Playwright seamlessly though.

**Comment by MindaugasMateika** [comment]
https://github.com/microsoft/playwright/issues/26507
to scale testing on our own without relying on CI resources

**Comment by richmar10** [comment]
In vscode, I'd like the ability to add timing instrumentation to our own packages so that when I'm running tests, I can tell where playwright is spending its time.  When the test setup process takes time, it's important to know how long it takes so that we don't end up extending test timeouts to unreasonable numbers.  Thx!

**Comment by luisecm** [comment]
Support for mobile and desktop 💃

**Comment by luisecm** [comment]
> Support for mobile and desktop 💃

Just kidding, but I wish there is something as good as playwright for desktop/mobile

**Comment by gjmargallo** [comment]
It would be nice if you could enrich the default HTML test reports with new options like pie charts etc.

Also to add native app testing, for instance, by using clouds based on appium. At the end, E2E testing requires most of the time web, api, app & db steps combined to validate a customer journey. If we cannot do some "parts" (e.g., app), Playwright cannot validate the whole customer journey, there is no workaround for this.

Our companies pay us as testers to cover & protect those flows.

**Comment by vitalets** [comment]
In 2025 I wish Playwright will integrate some AI stuff to help with writing and debugging e2e tests 🎄🎄🎄

**Comment by gjmargallo** [comment]
I wish a new AI locator function _getbyDescription()_, where I can enter something like _Button used to submit my form, labelled as Sign in_.

During run-time, the locator would connect to an AI service, which will determine in real-time the best locator for this _human_ description based on the current page DOM.

**Comment by erayerdin** [comment]
#11164 is definitely a missing feature. Firebase utilizes IndexedDB (for some weird reason) in order to save its authentication state. Working with it is a nightmare. It causes a lot of headaches like #34000.

**Comment by Hacktivist668** [comment]
#2903

**Comment by marcusNumminen** [comment]
I would like to see this one:
https://github.com/microsoft/playwright/issues/23354
(make it possible to run the retries in the end of the run)

**Comment by apatelia** [comment]
> > Support for mobile and desktop 💃
> 
> Just kidding, but I wish there is something as good as playwright for desktop/mobile

https://github.com/empirical-run/appwright

I haven't tried it yet, but it looks promising.

**Comment by kristaps-aksjonovs** [comment]
https://github.com/microsoft/playwright/issues/22937 -- This is blocking to use parallel runs :(

**Comment by StanislavGrishaev** [comment]
https://github.com/microsoft/playwright/issues/23827

**Comment by fpapado** [comment]
I think this fits with the holiday spirit: **I would like for Playwright to make the "View Trace" button more prominent in the HTML report**.

 As it stands, the test title link, the filename link and the video link all take the user to the `#?testId` page, but only the tiny "trace viewer" (rectangle sliced in three parts) icon link takes them to the trace viewer. There is no visible label for those icon links.

At work, we had to point folks to the trace viewer multiple times, because it is easy to miss. This is a bit of a shame, because the trace viewer is the best part of Playwright, in my view 😅 

This became annoying enough that we started overwriting the HTML reports and injecting some CSS to brighten the trace viewer button up ✨ It's just some CSS targeting `a[href^="trace/"]` and related parts (`svg` and `::after`), but it would be cool to see something built-in, or first-class extension point to the viewer.

(I can file an issue if you think this makes sense 😇)

![Image](https://github.com/user-attachments/assets/23ad00f9-0644-40cb-836f-5c80c6c619d3)
![Image](https://github.com/user-attachments/assets/e5ccc1df-7815-4889-a7ab-3c4824cd3a07)

**Comment by orizens** [comment]
### Report / Table View:
1. memorize collapsed tests (idea - via search query)

### Trace View Related:
1. filter / search network tab
2. highlight related network rows per selected action (in the Actions panel)
3. highlight / indicate `waitForResponse` in network tab
4. show "fill" values in Actions (locator now does)

**Comment by terradek** [comment]
https://github.com/microsoft/playwright/issues/28542
Currently I use this
```JavaScript
await this.page.getByText(deviceName).waitFor({ timeout: 200 }).catch(() => { });
```

**Comment by hardeep-qa** [comment]
Integration with Azure test plan.

**Comment by TusharAnandTR** [comment]
Please help with this:
[[Bug]: Test Profiles are not showing up VS Code Playwright for VS Code extension](https://github.com/microsoft/playwright/issues/32183)

**Comment by marcusNumminen** [comment]
Accessing the request object through the response objects to get the method and headers. This is really good for debugging purposes

https://github.com/microsoft/playwright/issues/16239

We utilize "rest clients" and there if we get back a non 2xx response we try to parse info from the response and print it to the log. Sometimes we have the same endpoints for different methods and it hard to know with one it is. To over come this I have a dirt hack to find the name of the calling method, that looks something like this: 
`const callerName = new Error().stack.split('at ')[2].split(' ')[0];`

**Comment by quldude** [comment]
Be able to use hyperlinks in annotations in html report.

**Comment by vhuseinova-msft** [comment]
It would be great to improve visual comparison (especially for component tests): https://github.com/microsoft/playwright/issues/24312

**Comment by Frederick-Hughes** [comment]
Closed as "out of scope" (controversial) but worth a shot #7292

**Comment by vladkrasn** [comment]
1. Find element by custom attribute or its substring
2. Get full href link instead of only the latter part of it

**Comment by IbrahimAni** [comment]
I wish playwright will support mobile testing. 😊🧑‍🎄

**Comment by la55u** [comment]
Official Fedora support instead of hacky docker based workarounds.

**Comment by tivac** [comment]
https://github.com/microsoft/playwright/issues/14572

It's really difficult to do interesting things around modules w/o being able to mock them in playwright.

**Comment by PramodKumarYadav** [comment]
# Dynamic sharding based on the test files to run in a test run.

Currently we have to predefine matrix strategy in GitHub and would have a fixed number of runners running tests irrespective of if there are less or more tests to run in that run. 

It would be awesome if playwright team can think of a solution to provide dynamic number of runners based on the number of test files to run. 

Say based on a ratio of files to run vs total number of test files. User decides max runners and then runners are assigned based on a product of ratio of files changed multiplied by total runners.

**Comment by yuri-pires** [comment]
No more Appium, Selenium, that's the past

We want Playwright for mobile testing on Android and IOS

**Comment by roniemartinez** [comment]
Rust support please.

**Comment by StevenBoutcher** [comment]
Electron support exiting Experimental mode 😉

**Comment by StevenBoutcher** [comment]
Viable alternative to XCUITest and Espresso

**Comment by StevenBoutcher** [comment]
your own education platform, a "Playwright University" with interactive exercises and shareable certificates

**Comment by christian-bromann** [comment]
WebDriver Bidi adoption and the ability to test in real browser!

**Comment by CloneOfAlex** [comment]
Please fix 700 defects - https://github.com/microsoft/playwright/issues

**Comment by nathanhannig** [comment]
Microsoft Playwright Testing 8n Azure Gov Cloud

Coverage reporter that can be used by SonarQube

**Comment by udhayak4251** [comment]
As no complaint right now, thanks for the amazing tool which leads to open source best web automation tool from my experience

**Comment by WallPetrucci** [comment]
> I wish playwright will support mobile testing. 😊🧑‍🎄

This is all I need for my stack to be fully Playwright

**Comment by Abishek7200** [comment]
AI test generator

**Comment by Airkro** [comment]
Store multiple snapshots in one (or more) file

**Comment by apis3445** [comment]
this #14813

**Comment by apis3445** [comment]
this #21484

**Comment by apis3445** [comment]
this #4870 I would like to include sound for some video testing and to check accessiblity

**Comment by apis3445** [comment]
It's not related to playwright but the return of this website to practice testing https://cloudtesting.contosotraders.com

**Comment by apis3445** [comment]
It's not related to playwright but better integration with azure devops pipeline. Now it's with external npm package alex_neo/playwright-azure-reporter  but I can't link to test cases as the same as .net that is linked in azure devops.

**Comment by apis3445** [comment]
the final version of  Microsoft Playwright Testing

**Comment by suresh-pandi** [comment]
support for native mobile applications (Android & IOS)

**Comment by DavitV-yan** [comment]
I would like to see Next and Previous buttons in the HTML reporter for navigating through the details of failed tests, now you have to go back then select the next test from the main reporter page

**Comment by DavitV-yan** [comment]
I would like a UI mode similar to the real-time mode in Cypress, instead of just generating screenshots from tests. This would be more useful when writing new tests, now it is immobility to use during the wiring new tests

**Comment by itsdouges** [comment]
- Electron support moving to stable
- Docs improved for showing how to run Playwright for VS Code extension authors
- Improvements for teardown exceeded timeout errors. Either debugging, forcing the nuclear option, or something else.

**Comment by craitose** [comment]
### **_Real iOS & Android Devices_**

Like others have mentioned, the ability to run tests on real mobile device browsers (iOS and android). Right now we have to rely on appium and webdriverio on a MacBook to achieve this. Emulation is just not the same.

**Comment by marcusNumminen** [comment]
I would like to have a command line interface in the codegen or in VS code when you record a test that exposes playwright object connected to the browser. This would be very handy to have in those cases when the codegen don't give you good locator or you want to try out different strategies.
I have done a quick hack for a simple CLI that I use a lot (a simplified version can be found here: https://github.com/marcusNumminen/playwrepl.ts)

For example when you use codegen for interacting with tables, In many cases this do not give you any good locator, Perhaps you want to try out different ways of using table headers or row names with a cli this could be a very easy task to play around to get a good locator for this.
Another example hard to reproduced front end issues. We had a front end issue that was hard to reproduce and with my cli I could just play around and test with different types of mocked data and some point a clicks in the browser to finally being able to reproduce the issue.

I would love to have this integrated cli functionality in the playwright echo system.

**Comment by Xotabu4** [comment]
Minor improvement: different number of workers per project

**Comment by grajesh690** [comment]
- Full Comprehensive API Testing
- Customizable HTML report something like Allure / Extent Reporter 
- Complete Electron Application support
- Mobile app testing
- Improved docs on wait strategies along with best practices
- Better tagging and grouping test cases 
- Isolated Setup.spec.ts based on the tests
- Improved documentation and best practices on environmental variables
- Separate project structure for BDD development 
- Inbuilt GenAI integrations

**Comment by Garretwar** [comment]
I need a function if i get  specific console or network error stop this test and do one more time, test shoud be flaky. 
Example CORS error

**Comment by lkasianenko-splash** [comment]
I want to enable parallel execution of multiple projects in Playwright, with a behavior based on worker count. Currently, when I set workers=2, the runner executes tests within the same project sequentially, such as project1 spec1 and project1 spec2.

Instead, I want the ability to run specs across different projects in parallel. For example, with workers=2, I would like project1 spec1 and project2 spec1 to execute concurrently. This would allow efficient distribution of tests across multiple projects.

**Comment by MaksymChepilDevPro** [comment]
This one: [[Feature] allow to NOT restart before and after hooks on test failure and rerun](https://github.com/microsoft/playwright/issues/14586)

**Comment by everton-dev** [comment]
I'd like to see  a fully integration with MS Dynamics 365 to facilitate the automation development.

**Comment by terrium88** [comment]
Hello. 
- Add support http 2.O to make api requests (APIRequest).  
- Add  report of tests for Java client.  https://playwright.dev/docs/api/class-reporter  for Java client does not support.

**Comment by son-qa** [comment]
Ability to add custom waits - selenium provides explicit wait

**Comment by ruifigueira** [comment]
Table cell locators by header: https://github.com/microsoft/playwright/issues/4646#issuecomment-2473131257

There is a [spec for establishing a relationship between data cells and header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), and screen readers implement a similar approach.

That would allow chaining locators to determine rows with specific header values , e.g.: row with `Paris` as `City` and `France` as `Country`.

**Comment by vitalets** [comment]
> This one: [[Feature] allow to NOT restart before and after hooks on test failure and rerun](https://github.com/microsoft/playwright/issues/14586)

And its sibling: [[Feature]: An option to NOT shutdown worker after test failure](https://github.com/microsoft/playwright/issues/32803)

**Comment by balazs-ita-epam** [comment]
Web UI mode for Python

**Comment by rcunniffe** [comment]
Ideally better documentation around using azure pipelines rather than GitHub :)

**Comment by Manuela-Sousa** [comment]
Integration with a model based testing approach, enabling the possibility of generating the tests automatically from the model also allowing the results to be integrated with another third party tools like test or project management tools.

**Comment by berkbltc** [comment]
Save screenshots using webp format ( #22984 )

**Comment by apis3445** [comment]
This #23586

**Comment by tetiana-kolos** [comment]
allow to copy any available text in the "Actions" section (in the Test Runner and Trace Viewer) - provides convenience and saves time

**Comment by cctui-dev** [comment]
https://playwright.dev/docs/test-sharding#merging-reports-from-multiple-shards

Updating merge-reports to support junit/xml
Also updating the documentation to say that it currently supports merging json would be a quick win

Thanks all and seasons greetings 🎅

**Comment by rorymichele** [comment]
similar to https://github.com/microsoft/playwright/issues/11164 I would love to see convenience methods to dump and restore indexedDB state. ideally this would be included in the exisiting `BrowserContext.storageState()` method, with an option to deactivate.

Restoring in the `BrowserContext` constructor is ok, but why not also have a `BrowserContext.restoreState()` method (that we access through `page.context()` rather than having to create the BrowserContext oneselves?

Thanks!

**Comment by Aghassi** [comment]
👋 I have a few requests from using Playwright at scale with Bazel that I'd like to cite here.

1. Give developers a way to turn off `transform` and other convenience features. At scale, we have tooling that does most of this work ahead of time so as to keep the test runners fast.
2. Provide a way to customize error messaging. Right now, I have to patch playwright to override certain outputs. When using tools like Bazel, I need to do things like update paths in the error messages to be stuff that Playwright can't determine
3. Allow me to orchestrate the watch mechanism through a custom function. Right now, watch uses the disk watcher, but this doesn't work well with Bazel when using tools like https://github.com/bazelbuild/bazel-watcher. I'm currently holding a patch that fires the watch mechanism when the bazel-watcher fires `stdin` information
4. Allow me to customize path resolution for source maps and the trace viewer. This goes back again to Bazel given that the path resolution by default will not find the intermediate artifacts that I'm generating. This again can be a function I pass into the config which you run on my behalf if it is provided.
5. I'd love a way to get memory and cpu pressure when tests flake or fail. I don't believe `--trace` covers this today, but having this would be very helpful to trying to determine root cause of flakiness.
6. Provide docs on bring your own browser for safari and firefox. I believe there is some patching going on today. Going back to using Bazel, I will be passing Playwright the browsers as I will fetch them outside of Playwright's purview. Having an understanding of how to apply the patches Playwright carries would be helpful.

Other than these specifics, we really enjoy using Playwright and are looking forward to the future improvements in the pipeline.

**Comment by chebmohamed** [comment]
I’d like to suggest adding a structure like BDD (for example, "Given-When-Then") to Playwright. This would make it easier to create test scenarios with end users and use them as documentation.

**Comment by gkiely** [comment]
`test.todo()`

**Comment by boryskrutskykh** [comment]
I'd like to have feature. Mobile testing Android and iOS.

**Comment by Kranael** [comment]
Websocket testing and generally more/easier api testing capabilitys!

**Comment by tphillips-ict** [comment]
> Websocket testing and generally more/easier api testing capabilitys!

Playwright does support websockets! https://playwright.dev/docs/network#websockets

**Comment by omarchouikh** [comment]
Playwright Lighthouse Core Web Vitals<3

**Comment by Kranael** [comment]
> > Websocket testing and generally more/easier api testing capabilitys!
> 
> Playwright does support websockets! https://playwright.dev/docs/network#websockets

I know I mean it at API testing level no page. So we need something like a ws fixture like we have with request baked in

**Comment by ilorwork** [comment]
Simplify maximizing browser window.

**Comment by Az8th** [comment]
Extend client certificates support to add PKCS#11 would be wonderful 🤞
We would be able to automate scenarios that includes physical cryptographic devices interaction

**Comment by pmicko** [comment]
As a long-time WebStorm user, I would love the ability to open the related test file in my IDE by clicking the 'Show source' button in UI mode 🚀  

Currently, the button only works for opening files in VS Code.

There is already an issue filed for this: https://github.com/microsoft/playwright/issues/22158

**Comment by ukco** [comment]
#8208 
don't ignore electron anymore, pls

**Comment by angelo-loria** [comment]
Please don't add AI garbage. I'd like to see the ability to filter test runs by repo on the Microsoft Playwright Testing dashboard and an easier way to record videos with multi-page tests/page objects in fixtures.

**Comment by Darkensses** [comment]
An undetected chrome feature, just like https://github.com/seleniumbase/SeleniumBase does 👀🐱‍👤

...and search params page assertion #32708

**Comment by hernaanm** [comment]
Eager to see some Server Side mocking since its really a pain for SSR applications 

already created -> https://github.com/microsoft/playwright/issues/30766

**Comment by mastrzyz** [comment]
Ability to understand the worker failures better

![Image](https://github.com/user-attachments/assets/2f79f2d0-639c-4927-a519-44a2d3ddf19a)

**Comment by Diogooliveira-tech** [comment]
> Tell us what you'd like to see in Playwright in the next year!

I would like more features to be implemented and the documentation for .NET (C#) to be improved, especially with features similar to those available in Node.js, which are easier to use.

**Comment by diginikkari** [comment]
First of all I want to thank you for all the hard work you have done with the playwright ❤️

It’s hard to think what to wish for because you seem to be a head of me and bring stuff I didn’t dare to wish for. 

What I know is somehow possible, but not have achieved myself is to have guidance how to configure code coverage using playwright, vite and chromium. Have a merry holidays you all 🎄

**Comment by SofianeH1** [comment]
Add tests with update snapshots to the -onlyChaged test run

**Comment by NoamGaash** [comment]
Adding some advanced HAR configuration would be cool (see #21405). Also, seems like "minimal" mode doesn't work well.

Also, it could be nice if `page.clock` had an API to freeze the datetime, and resolve timeouts imminently. It would be useful for tooltips and autocomplete components that has denounce in the entrance animation

**Comment by marnandr** [comment]
I have three ideas that I think would be super helpful:
1. Official support for Azure Test Plans ⚗ reporting (#14839).
2. Improved support for Azure Pipelines 🚀 (I will do my best to contribute to the documentation).
3. Support for print dialog events 🖨 as `page.waitForEvent()` (#6543).

**Comment by lkasianenko-splash** [comment]
Want to have the ability run describes in serial mode but tests inside in default mode like for parallel


```test.describe.configure({ mode: 'serial' });

test.describe('A, runs in parallel with B', () => {
  test.describe.configure({ mode: 'default' });
  test('in order A1', async ({ }) => { expect(true).toBe(true); });
});

test.describe('B, runs in parallel with A', () => {
  test.describe.configure({ mode: 'default' });
  test('in order B1', async ({ }) => { expect(true).toBe(false); });
  test('in order B2', async ({ }) => { expect(true).toBe(true); });
});```

**Comment by cctui-dev** [comment]
I'd also love `test.step.skip()` to be implemented :)

**Comment by olaf-k** [comment]
an integrated, no-nonsense, reliable and simple to use coverage gathering/reporting mechanism. keep up the excellent work and merry xmas people!

**Comment by ramam-v** [comment]
Native BDD support please

**Comment by tdhanjun** [comment]
self-cure

**Comment by muhqu** [comment]
Would also love to see custom asymmetric matchers to be fully supported...
* https://github.com/microsoft/playwright/issues/32562

**Comment by yonatanai** [comment]
Option to extend the HTML reporter or easily create custom UI reporter using some common functionality with HTML reporter like iterating steps and generating the code, etc.

**Comment by Bilge** [comment]
Considering a typical workflow comprises using codegen to generate (bad) selectors for various parts of a page and then jumping into an editor to fine-tune all the selectors to make them more robust, it would be nice if the codegen macro tool itself presented alternatives to choose from when building locator paths.

**Comment by Wish-Infinite** [comment]
Appium like support for Mobile Native App automation.

**Comment by alexanderadam** [comment]
Support for 🦊  Firefox 🧩 extensions, which is requested every now and then:

- #2644
- #7297
- #9202
- #15299
- #16544

**Comment by Jaakkonen** [comment]
Running non-patched Firefox builds with BiDi support :) Also should fix the Firefox extension issue
* https://github.com/microsoft/playwright/issues/32577

**Comment by AlaaEl-DinAhmed** [comment]
Component Testing for Angular

**Comment by fMoro1999** [comment]
> Component Testing for Angular

Yes, please!!!

**Comment by shilCode** [comment]
> Component Testing for Angular

Pretty please!

**Comment by pavelfeldman** [comment]
Thank you for your requests! Santa has noted all your requests and returned to the North Pole to decide which ones to do first! Happy New Year!

**Comment by wangdatu** [comment]
this https://github.com/microsoft/playwright/pull/31823

**Comment by hae-ctr-leonid-vrublevskyi** [comment]
Add ability to work with desktop/mobile  app.

**Comment by lvrub** [comment]
Add ability to work with desktop /mobile apps.

**Comment by radekBednarik** [comment]
neovim plugin for Playwright ;) :P

**Comment by robbiewareham** [comment]
Rich logic experience around when to capture response bodies in a HAR file.

Example:

**Status Code**

I want the response body for a static resource request when the status code is >= 400

**URL Reg Ex**

The AUT may have XHR requests that are mundane but large and i am not interested in them, but some XHR are crucial and i want to always capture response bodies


---

I am going to implement this post-processing but to have this defined before the test starts would be awesome

**Comment by mandras73** [comment]
Playwrigh Pytest specific ambassadors and release note presentations

**Comment by dylannnn** [comment]
Vote for Component Testing with Angular

**Comment by TomasHubelbauer** [comment]
Please recognize the semantic difference between `test.todo` and `test.fixme` and introduce the former (there are already several closed issues for this which could be reopened) or codify the team's view they aren't semantically different to them (which I don't agree with, but it's ultimately not my call) by introducing a `test.todo` alias for `test.fixme`.
